web-admin/src/features/scheduled-reports/query-mapper.ts

```diff
@@ -0,0 +1,277 @@
+import { getSortType } from "@rilldata/web-common/features/dashboards/leaderboard/leaderboard-utils";
+import { SortDirection } from "@rilldata/web-common/features/dashboards/proto-state/derived-types";
+import { getProtoFromDashboardState } from "@rilldata/web-common/features/dashboards/proto-state/toProto";
+import { useMetaQuery } from "@rilldata/web-common/features/dashboards/selectors";
+import { getDefaultMetricsExplorerEntity } from "@rilldata/web-common/features/dashboards/stores/dashboard-store-defaults";
+import type { MetricsExplorerEntity } from "@rilldata/web-common/features/dashboards/stores/metrics-explorer-entity";
+import { initLocalUserPreferenceStore } from "@rilldata/web-common/features/dashboards/user-preferences";
+import { isoDurationToFullTimeRange } from "@rilldata/web-common/lib/time/ranges/iso-ranges";
+import {
+  type DashboardTimeControls,
+  TimeRangePreset,
+} from "@rilldata/web-common/lib/time/types";
+import {
+  createQueryServiceMetricsViewTimeRange,
+  V1TimeRange,
+  type V1MetricsViewAggregationRequest,
+  type V1MetricsViewComparisonRequest,
+  type V1MetricsViewRowsRequest,
+  type V1MetricsViewSpec,
+  type V1MetricsViewTimeSeriesRequest,
+  type V1MetricsViewToplistRequest,
+  type V1Resource,
+  type V1TimeRangeSummary,
+} from "@rilldata/web-common/runtime-client";
+import { runtime } from "@rilldata/web-common/runtime-client/runtime-store";
+import { derived, get, type Readable, readable } from "svelte/store";
+
+type ReportQueryRequest =
+  | V1MetricsViewAggregationRequest
+  | V1MetricsViewToplistRequest
+  | V1MetricsViewRowsRequest
+  | V1MetricsViewTimeSeriesRequest
+  | V1MetricsViewComparisonRequest;
+
+type QueryMapperArgs<R extends ReportQueryRequest> = {
+  dashboard: MetricsExplorerEntity;
+  req: R;
+  metricsView: V1MetricsViewSpec;
+  timeRangeSummary: V1TimeRangeSummary;
+  executionTime: string;
+};
+
+type ParsedReportQuery = {
+  ready: boolean;
+  state?: string;
+  metricsView?: string;
+  error?: string;
+};
```

Comment:
I'm noticing we have a lot of places in the codebase where we use a derived store to combine the results from two queries. Rather than a type like this one that mixes query status (`ready`, `error`) with query results (`state`, `metricsView`), it could be good if we created a generic type to hold the query status, like...

```typescript
type BaseQuery<T> = {
  isReady: boolean;
  error?: string
  data: T
}
```

which we could use to wrap the query results type...

```typescript
type DashboardState = {
    metricsView: string
    state: string
}

type DashboardStateQuery = BaseQuery<DashboardState>;
```

Reply:
FWIW TanStack Query v5 will give us a first-class way to combine multiple queries, and the above approach mirrors TanStack.

Reply:
Good point. I will not refactor other places but update this derived state.